### PR TITLE
Support `LightClientFinalityUpdate` and `LightClientOptimisticUpdate` rpcs

### PR DESCRIPTION
**Pull request by GeemoCandama**
## Issue Addressed

partially addresses #3651.

## Proposed Changes

- Support `LightClientFinalityUpdate` https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#getlightclientfinalityupdate
- Support `LightClientOptimisticUpdate` https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#getlightclientoptimisticupdate

### CODE REVIEW

File: beacon_node/lighthouse_network/src/peer_manager/mod.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/peer_manager/mod.rs]: Just to confirm, these rpc requests will be only from light clients and not servers, so it doesn't make sense that they rate limit us?

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: Nothing needs to be encoded here since we have no request content. Can just return (similar to `MetaData`)

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: same as above

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: The request here can be empty. Basically we do not need a `LightClientOptimisticUpdateRequest` type at all here.

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: same as above.

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: As a light client server, we will never be making any Light client related rpc requests so we should not be getting responses for these messages. However, this could potentially be useful in testing (not sure how exactly since we don't support the outbound requests for light client rpcs at the libp2p level).
I'm not sure if we should handle this case or if we should just leave a comment here and return an error
@divagant-martian thoughts?

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: same as above.

File: beacon_node/lighthouse_network/src/rpc/methods.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/methods.rs]: I think we don't need these types at all since the protocol does not expect any request content.

File: beacon_node/lighthouse_network/src/rpc/methods.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/methods.rs]: same as above.

File: beacon_node/lighthouse_network/src/rpc/outbound.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/outbound.rs]: As mentioned, this need not have any internal type at all.

File: beacon_node/lighthouse_network/src/rpc/outbound.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/outbound.rs]: same as above.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: These should be just `RpcLimits::new(0,0)`.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: same as above.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: this should be the reponse limit, not request.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: same as above.

**Review by pawanjay176 — DISMISSED** [review]
Looking good. Just a couple of changes.
Also looks like we were returning `false` for `has_context_bytes` for all light client rpcs. This fixes the issue https://github.com/sigp/lighthouse/pull/3972 . Something that we would need to pull into this PR.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (GeemoCandama) [reply beacon_node/lighthouse_network/src/rpc/protocol.rs]: Should I change the LightClientBootstrap one as well?

File: beacon_node/lighthouse_network/src/peer_manager/mod.rs
Comment (GeemoCandama) [reply beacon_node/lighthouse_network/src/peer_manager/mod.rs]: Yes you are correct, so should we just return instead of `PeerAction::LowToleranceError` or something else? I fixed the other stuff other than merging in realbigsean's branch and adding the case for this the new rpcs. That should be easy but I've heard that it will make it harder to review if I rebase which I'll need to do in order to merge.

**Comment by divagant-martian** [comment]
I discussed this with @pawanjay176 and we might want to split the protocol modeling for asymmetric protocols before including these, FYI. Pawan is looking into this. If he doesn't have the bandwidth I might get to this at some point

**Comment by GeemoCandama** [comment]
Should be good now. I was trying to keep the branch up to date with the changes and I didn't realize that the changes broke it. Also I added a full match to the context_bytes function so it will give more compile time feedback, since I didn't realize that I needed to make those changes until after a bit of hunting.

**Comment by divagant-martian** [comment]
there are some conflicts. Would you mind looking into those and then I'll trigger ci for you

**Comment by GeemoCandama** [comment]
conflicts are fixed now.
EDIT: Clippy was the only that failed and I think I appeased clippy now, but when I ran clippy locally last time I didnt get any errors either so I am not certain.

**Comment by pawanjay176** [comment]
Hi @GeemoCandama , I had a look at this over the weekend. It's looking good. I'm gonna spend some time this week to see if the `MockRpcHandler` can be simplified. Sorry for the delay.

**Comment by jimmygchen** [comment]
Thanks @GeemoCandama for your work on this! I'm going to pick this one up and try to push it over the finishing line!

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: I'm not sure if there's any good reason to try to decode this as we don't make outbound requests - it would be an unexpected scenario, and the response isn't going to be useful for us I think. I'll make this return an error and adda comment.

File: beacon_node/lighthouse_network/src/peer_manager/mod.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/peer_manager/mod.rs]: Yeah I don't think we should ban them in this case (although it's an unexpected scenario) - I'll make it return and add a comment. This may change in the future if we decide to support running an actual light client using Lighthouse.

**Review by dapplion — COMMENTED** [review]
The network wiring of this PR is correct, but half of the diff is a mock to test the light-client network server. Instead I would suggest to implement the network consumer part in light-house network. This would prevent having code too coupled into libp2p mechanics that needs explicit maintenance + there's a good usecase for the consumer: use LC protocol to trustlessly acquire a recent finalized block root and then checkpoint sync.

There's no mechanism to trigger update publishing, so this PR won't add funtionality to unstable. I would suggest to put this PR on ice, develop a progressive lightclient server (eagerly produce messages) which can trigger gossip publishing. Then wire the network off these commits.

**Comment by jimmygchen** [comment]
> The network wiring of this PR is correct, but half of the diff is a mock to test the light-client network server. ..

Yep agree with the above, also mentioned your suggestion to @michaelsproul last week - we think it's a good idea.
The current `mock-light-client` pretty much need a re-write because of the `rust-libp2p` upgrade - I've made an unsuccessful attempt a few weeks ago, and that got me into asking myself the same question on whether we should come up with an alternative - I then started building a light client consumer prototype for this, but haven't got it to a usable state.

**Comment by jimmygchen** [comment]
Blocked by #4926

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: We think it may be useful to handle outbound requests for testing, will cover this as part of #4940

**Comment by jimmygchen** [comment]
I've made some updates to this branch and I think it's in a good state for review. I'd probably wait until we get #4946 merged in first, though.

We should also add some tests for this (#4940), I'll get to it when I have some time.

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: This needs to be updated to match the `LightClientHeader` variant.

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: Update rpc limits as this is no longer fixed length

File: beacon_node/lighthouse_network/src/service/api_types.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/service/api_types.rs:77]: Note this may change in the future for testing purposes.

**Review by jimmygchen — COMMENTED** [review]
I've re-reviewed this PR, there are a few things that needs to be updated once we merged #4896, and we can merge this one next.

**Review by dapplion — APPROVED** [review]
Code is solid! Glad to this moving 🔥  since it uses the cached data it does not expose the node to a significant load when enabled

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/rpc/protocol.rs]: https://github.com/sigp/lighthouse/pull/3849/commits/c888923304cc5dbed7d42b9cc0a4addd821ecdef

**Comment by jimmygchen** [comment]
I've merged in latest `unstable`. I haven't been able to test this in a real network as we still haven't implemented the `LightClientUpdates` req/resp interface. I think we can merge this as it is after review, as it doesn't touch the default code path (requires `light-client-server` flag).

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (dapplion) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs]: Why not default to Deneb for now?

Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/rpc/protocol.rs]: I'm not familiar with the Electra changes to light clients, and I'd rather not forget to update this when Electra comes. What's the value of adding a value that might be obsolete? We're not going to hit this code path pre Electra.

Comment (dapplion) [reply beacon_node/lighthouse_network/src/rpc/protocol.rs]: I worry we forget, enable the light-client server by default and LH gets panics on devnets / testnets. Could add a default to electra and a `TODO(electra)`

**Review by dapplion — APPROVED** [review]
Let's merge this! We can address the Electra data structures latter 👍

File: beacon_node/lighthouse_network/src/rpc/protocol.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/rpc/protocol.rs]: Done in 51fe0ce83cb3eeabee4e367424bb9aa20e91537c

**Review by dapplion — APPROVED** [review]

**Comment by jimmygchen** [comment]
@mergify queue

**Pull request merged**
